client/tracing/src/lib.rs

```diff
@@ -290,41 +309,65 @@ impl Subscriber for ProfilingSubscriber {
 			overall_time: ZERO_DURATION,
 			values,
 		};
-		self.span_data.lock().insert(id, span_datum);
-		Id::from_u64(id)
+		{
+			let mut span_data = self.span_data.lock();
+			if span_data.len() > SPAN_LIMIT {
+				log::warn!("Accumulated too many spans, discarding oldest");
+				let mut keys = span_data.keys().map(|id| id.into_u64()).collect::<Vec<_>>();
+				keys.sort();
+				for key in keys[0.. SPAN_LIMIT / 10].iter() {
+					span_data.remove(&Id::from_u64(*key));
+				}
+			}
+			span_data.insert(id.clone(), span_datum);
+		}
+		id
 	}
 
 	fn record(&self, span: &Id, values: &Record<'_>) {
 		let mut span_data = self.span_data.lock();
-		if let Some(s) = span_data.get_mut(&span.into_u64()) {
+		if let Some(s) = span_data.get_mut(span) {
 			values.record(&mut s.values);
 		}
 	}
 
 	fn record_follows_from(&self, _span: &Id, _follows: &Id) {}
 
-	fn event(&self, _event: &Event<'_>) {}
+	fn event(&self, event: &Event<'_>) {
+		let mut values = Values::default();
+		event.record(&mut values);
+		let trace_event = TraceEvent {
+			name: event.metadata().name(),
+			target: event.metadata().target().to_owned(),
+			level: event.metadata().level().clone(),
+			values,
+			parent_id: event.parent().cloned().or_else(|| self.current_span.id()),
+		};
+		self.trace_handler.handle_event(trace_event);
+	}
 
 	fn enter(&self, span: &Id) {
+		self.current_span.enter(span.clone());
 		let mut span_data = self.span_data.lock();
 		let start_time = Instant::now();
-		if let Some(mut s) = span_data.get_mut(&span.into_u64()) {
+		if let Some(mut s) = span_data.get_mut(&span) {
 			s.start_time = start_time;
 		}
 	}
 
 	fn exit(&self, span: &Id) {
+		self.current_span.exit();
```

Comment:
I am not a 100% convinced this is the correct API... Why would we have an `enter` _with_ the id, but then when exit wouldn't need that `id` – especially since the trait otherwise requires them? That seems like we might be using this wrong. Like, are we sure we are keeping the "current thread" information accurate here? Just want to make sure that we don't accidentally enter and exit on other entries and don't notice...

Reply:
Looking at the docs this appears to be the correct way to use it - it's just a stack push/pop; in a follow-up PR this responsibility should be handled the FmtSubscriber anyway.